File: packages/domain/test/practice-catalog.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { readFileSync } from 'node:fs';
import { practiceCatalogActivityIds, practiceCatalogById, practiceCatalogContentVersion,
  practiceCatalogFloors, practiceCatalogPayloadVersion, practiceCatalogSchemaVersion } from '../src/practice-catalog.generated.js';

test('compiled catalog retains every field from the versioned JSON and is deeply immutable', () => {
  const source = JSON.parse(readFileSync(new URL('../../../content/practice-catalog.json', import.meta.url), 'utf8')) as {
    schemaVersion: number; contentVersion: string; payloadVersion: number;
    floors: {number: number; title: string}[];
    activities: {id: string; floor: number; title: string; correctChoiceId: string;
      acceptedChoiceIds: string[]; choiceIds: string[]; introducedIn: number; prerequisiteId: string | null}[];
  };
  assert.equal(practiceCatalogSchemaVersion, source.schemaVersion);
  assert.equal(practiceCatalogContentVersion, source.contentVersion);
  assert.equal(practiceCatalogPayloadVersion, source.payloadVersion);
  assert.deepEqual(practiceCatalogFloors, Object.fromEntries(source.floors.map(floor => [floor.number, floor.title])));
  assert.deepEqual(practiceCatalogActivityIds, source.activities.map(entry => entry.id));
  assert.deepEqual(practiceCatalogActivityIds.map(id => practiceCatalogById[id]), source.activities);
  assert.ok(Object.isFrozen(practiceCatalogById));
  assert.ok(Object.isFrozen(practiceCatalogActivityIds));
  for (const entry of Object.values(practiceCatalogById)) {
    assert.ok(Object.isFrozen(entry)); assert.ok(Object.isFrozen(entry.choiceIds));
    assert.ok(Object.isFrozen(entry.acceptedChoiceIds));
    assert.throws(() => Object.assign(entry, {title: 'Mutated release'}), TypeError);
  }
});

File: apps/mobile/test/practice_catalog_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:trimmy/design_study/activities.dart';
import 'package:trimmy/design_study/practice_catalog.generated.dart';

void main() {
  test('generated Dart metadata matches every canonical JSON field', () {
    final source =
        jsonDecode(
              File('../../content/practice-catalog.json').readAsStringSync(),
            )
            as Map<String, dynamic>;
    expect(practiceCatalogSchemaVersion, source['schemaVersion']);
    expect(practiceCatalogContentVersion, source['contentVersion']);
    expect(practiceCatalogPayloadVersion, source['payloadVersion']);
    final entries = source['activities'] as List<dynamic>;
    expect(practiceCatalogActivityIds, entries.map((entry) => entry['id']));
    for (final dynamic entry in entries) {
      final generated = practiceCatalogById[entry['id']]!;
      expect(generated.id, entry['id']);
      expect(generated.floor, entry['floor']);
      expect(generated.title, entry['title']);
      expect(generated.correctChoiceId, entry['correctChoiceId']);
      expect(generated.acceptedChoiceIds, entry['acceptedChoiceIds']);
      expect(generated.choiceIds, entry['choiceIds']);
      expect(generated.introducedIn, entry['introducedIn']);
      expect(generated.prerequisiteId, entry['prerequisiteId']);
    }
    for (final dynamic floor in source['floors'] as List<dynamic>) {
      expect(practiceCatalogFloors[floor['number']], floor['title']);
    }
    expect(
      () => practiceCatalogActivityIds.add('changed'),
      throwsUnsupportedError,
    );
    expect(
      () => practiceCatalogById['check-the-date']!.choiceIds.add('changed'),
      throwsUnsupportedError,
    );
    expect(
      () => practiceCatalogById['review-team-update']!.acceptedChoiceIds.add(
        'changed',
      ),
      throwsUnsupportedError,
    );
  });

  test(
    'authored native activity IDs and answers agree with generated metadata',
    () {
      for (final activity in studyActivities) {
        final metadata = practiceCatalogById[activity.id];
        expect(metadata, isNotNull, reason: activity.id);
        expect(metadata!.title, activity.title);
        expect(metadata.correctChoiceId, activity.correctChoiceId);
        expect(metadata.acceptedChoiceIds, activity.effectiveAcceptedChoiceIds);
        expect(metadata.choiceIds, activity.choices.map((choice) => choice.id));
      }
    },
  );
}
